GET data stream API returns additional information

andreidan opened the pull request:
This adds the data stream's index template, the configured ILM policy
(if any) and the health status of the data stream to the `GET _data_stream`
response.

Relates to #58316 
Relates to #53100 

The proposed response looks like this:
```
{
  "data_streams" : [
    {
      "name" : "timeseries",
      "timestamp_field" : {
        "name" : "@timestamp",
        "mapping" : {
          "type" : "date"
        }
      },
      "indices" : [
        {
          "index_name" : ".ds-timeseries-000001",
          "index_uuid" : "m3Id6-T7TuOBuC3qoltdqQ"
        }
      ],
      "generation" : 1,
      "status" : "YELLOW",
      "template" : "timeseries_template",
      "ilm_policy" : "timeseries_policy"
    },
    {
      "name" : "timeseries_3",
      "timestamp_field" : {
        "name" : "@timestamp",
        "mapping" : {
          "type" : "date"
        }
      },
      "indices" : [
        {
          "index_name" : ".ds-timeseries_3-000001",
          "index_uuid" : "YJKrb9_yR2iukhi0vCy1zQ"
        }
      ],
      "generation" : 1,
      "status" : "GREEN",
      "template" : "timeseries_template",
      "ilm_policy" : "timeseries_policy"
    }
  ]
}
```

andreidan committed:
GET data stream API returns additional information

This adds the data stream's index template, the configured ILM policy
(if any) and the health status of the data stream to the GET _data_stream
response.

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

andreidan committed:
Remove unused import

andreidan committed:
Fix yaml tests

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into get_data_stream

andreidan committed:
Fix yaml test

andreidan commented:
Added WIP as the HLRC still needs some work

danhermann commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.data_stream/10_basic.yml:
```diff
@@ -49,16 +49,16 @@ setup:
-  - match: { 0.generation: 1 }
-  - length: { 1.indices: 1 }
-  - match: { 1.indices.0.index_name: '.ds-simple-data-stream2-000001' }
+  - match: { data_streams.0.name: simple-data-stream1 }
+  - match: { data_streams.0.timestamp_field.name: '@timestamp' }
+  - match: { data_streams.0.generation: 1 }
+  - length: { data_streams.0.indices: 1 }
+  - match: { data_streams.0.indices.0.index_name: '.ds-simple-data-stream1-000001' }
+  - match: { data_streams.1.name: simple-data-stream2 }
+  - match: { data_streams.1.timestamp_field.name: '@timestamp2' }
+  - match: { data_streams.0.generation: 1 }
+  - length: { data_streams.1.indices: 1 }
+  - match: { data_streams.1.indices.0.index_name: '.ds-simple-data-stream2-000001' }
```
I think that at least one of these REST tests should be updated with the full list of fields that are returned from the get data stream API.

andreidan replied:
Thanks Dan, pushed a match for `status` and `template` (the lifecycle policy is tested in `TimeseriesDataStreamIT` as it's an xpack setting)

danhermann reviewed (approved):
I left one comment about REST tests that I think should be addressed. LGTM, otherwise.

Edit: Just saw you marked it WIP. I'll take another look at it when you've completed it.

andreidan committed:
Adjust HLRC to new API

andreidan committed:
Parse ilm_policy as optional

andreidan committed:
Yaml test: match corresponding DS status and template

danhermann reviewed (approved):
LGTM. Thanks, @andreidan!

andreidan committed:
Fix stack yaml tests

andreidan committed:
Make the index template optional for the GET _data_stream response

Restoring a data stream from a snapshot could install a data stream that
doesn't match any composable templates. This makes the index template field
in the `GET _data_stream` response optional.

martijnvg commented on server/src/main/java/org/elasticsearch/action/admin/indices/datastream/GetDataStreamAction.java:
```diff
@@ -119,11 +214,13 @@ public void writeTo(StreamOutput out) throws IOException {
         public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-            builder.startArray();
-            for (DataStream dataStream : dataStreams) {
+            builder.startObject();
```
why was this top level `data_streams` field introduced? I don't think that this is needed?

danhermann replied:
It did make it easier to verify in REST tests that the number of returned data streams was as expected with a `- length: { data_streams: 0 }` directive. I don't think the length directive otherwise provides a way to refer to the root level of a response.

martijnvg replied:
I think the `- length: ` directive should be improved to handle this? Now the response format is changed, because a test directive can't handle something. 

I also don't see length directive being used on the `data_streams` object, so maybe just keep using:

```
- match: { $body: [] }
``` 

?

danhermann replied:
Those are fair points. 👍

andreidan replied:
Went of the top array field to be consistent with component and composable templates eg. https://github.com/elastic/elasticsearch/blob/35159090b943d1044929d29464ff9fa718d1700d/server/src/main/java/org/elasticsearch/action/admin/indices/template/get/GetComponentTemplateAction.java#L146
